microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md

```diff
@@ -0,0 +1,314 @@
+* `snapshot.meta.node_fields` to know how many fields each node has.
+
+Each node in the array is represented by a series of `snapshot.meta.node_fields.length` numbers. So the total number of elements in the `nodes` array is `snapshot.node_count` multiplied by `snapshot.meta.node_fields.length`.
+
+In order to recreate a node you need to read the numbers from the `nodes` array by groups of `snapshot.meta.node_fields.length`.
+
+Below is a code snippet that shows the node fields metadata and the data for the two first nodes in the graph.
+
+```json
+{
+    "snapshot": {
+        "meta": {
+            "node_fields": [
+                "type",
+                "name",
+                "id",
+                "self_size",
+                "edge_count",
+                "trace_node_id",
+                "detachedness"
+            ]
+            ...
+        }
+        ...
+    },
+    "nodes": [
+        9,1,1,0,10,0,0,
+        2,1,79,12,1,0,0,
+        ...
+    ]
+    ...
+}
+```
+
+| Index in node group | Name | Description |
+| --- | --- | --- |
+| `0` | `type` | The type of node. See [Node types](#node-types) to find out what are the possible types. |
+| `1` | `name` | The name of the node. This number corresponds to the index in the `strings` top-level array which you need to look-up to find the actual name. |
+| `2` | `id` | The node's unique ID. |
+| `3` | `self_size` | The node's size in bytes. |
+| `4` | `edge_count` | The number of edges connected to this node. |
+| `5` | `trace_node_id` | _Not documented yet_ |
+| `6` | `detachedness` | Whether this node can be reached from the `window` global object. `0` if it can be reached, `1` if it can't. |
+
+### Node types
+
+The first number found in the group of numbers for a node in the `nodes` array corresponds to its type. 
+
+This number is an index that can be used to lookup the type name in the `snapshot.meta.node_types[0]` array.
+
+| Node type | Description |
+| --- | --- |
+| Hidden | A V8 internal element that doesn't correspond directly to a user-controllable JavaScript object. In DevTools, these all show up under the category name **(system)**. Even though these objects are internal, they can be an important part of retainer paths. |
+| Object | Any user-defined object like `{ x: 2 }` or `new Foo(4)`. Contexts, which show up in DevTools as **system / Context** hold variables that had to be allocated on the heap because they're used by a nested function. |
+| Native | Things allocated by the Blink rendering engine, not by V8. These are mostly DOM: HTMLDivElement, CSSStyleRule, etc. |
+| Concatenated string | The result of concatenating two strings with the `+` operator. Rather than making a new string with a copy of all the data from the two source strings, V8 creates a ConsString object with pointers to the two source strings. From a JavaScript perspective, it acts just like any other string, but from a memory profiling perspective, it is different. |
```

Comment:
```suggestion
| Concatenated string | The result of concatenating two strings with the `+` operator. Rather than making a new string that contains a copy of all the data from the two source strings, V8 creates a `ConsString` object that contains pointers to the two source strings. From a JavaScript perspective, it acts just like any other string, but from a memory profiling perspective, it is different. |
```
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/u/using-by-using-with